File: frontend/src/utils/storage.ts
export type HasAuth = {
  twitter: boolean
}

export const cookies = () =>
  document.cookie.split('; ').reduce(
    (prev, current) => {
      const [name, value] = current.split('=')
      prev[name] = value
      return prev
    },
    {} as Record<string, string>
  )

export const setCookie = (
  name: string,
  value: string,
  expirationMillis: number | null = null
) => {
  let expires = ''
  if (expirationMillis) {
    const date = new Date()
    date.setTime(date.getTime() + expirationMillis)
    expires = `;expires=${date.toUTCString()}`
  }
  document.cookie = `${name}=${value}${expires};path=/`
}

export const clearCookie = (name: string) => {
  document.cookie = `${name}=;expires=${new Date(0).toUTCString()};path=/`
}

export const getJwtToken = (): string | undefined => cookies()['access_token']

export const clearJwtToken = () => clearCookie('access_token')

export const setJwtToken = (token: string) =>
  setCookie('access_token', token, 1000 * 60 * 60 * 24 * 2)

export const hasJwtToken = () => !!getJwtToken()

export const storeObjectInLocalStorage = <A>(key: string, value: A | undefined | null) => {
  if (value === undefined || value === null)
    clearObjectFromLocalStorage(key)
  else
    localStorage.setItem(key, JSON.stringify(value))
}

export const getObjectFromLocalStorage = <A>(key: string): A | null =>
  JSON.parse(localStorage.getItem(key) || 'null')

export const clearObjectFromLocalStorage = (key: string) =>
  localStorage.removeItem(key)

export const setAuthStatus = (hasAuth: HasAuth | undefined | null) =>
  storeObjectInLocalStorage('hasAuth', hasAuth)

export const getAuthStatus = (): HasAuth => {
  const hasAuth = getObjectFromLocalStorage('hasAuth') as HasAuth | null
  return hasAuth || { twitter: false }
}

export const updateAuthStatus = (f: (current: HasAuth) => HasAuth) => {
  const current = getAuthStatus()
  setAuthStatus(f(current))
}

File: frontend-kotlin/src/jsMain/kotlin/com/alexn/socialpublish/frontend/utils/Storage.kt
package com.alexn.socialpublish.frontend.utils

import kotlin.js.Date
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import web.dom.document
import web.storage.localStorage


@Serializable
data class HasAuth(
    val twitter: Boolean = false,
)

fun cookies(): Map<String, String> {
    val rawCookie = document.cookie
    if (rawCookie.isBlank()) {
        return emptyMap()
    }
    return rawCookie.split("; ").associate { entry ->
        val parts = entry.split("=", limit = 2)
        val name = parts.first()
        val value = parts.getOrNull(1) ?: ""
        name to value
    }
}

fun setCookie(name: String, value: String, expirationMillis: Int? = null) {
    val expires = if (expirationMillis != null) {
        val date = Date(Date().getTime() + expirationMillis)
        ";expires=${date.toUTCString()}"
    } else {
        ""
    }
    document.cookie = "$name=$value$expires;path=/"
}

fun clearCookie(name: String) {
    val date = Date(0)
    document.cookie = "$name=;expires=${date.toUTCString()};path=/"
}

fun getJwtToken(): String? = cookies()["access_token"]

fun clearJwtToken() = clearCookie("access_token")

fun setJwtToken(token: String) = setCookie("access_token", token, 1000 * 60 * 60 * 24 * 2)

fun hasJwtToken(): Boolean = getJwtToken() != null

fun setAuthStatus(hasAuth: HasAuth?) {
    if (hasAuth == null) {
        localStorage.removeItem("hasAuth")
    } else {
        val json = Json.encodeToString(hasAuth)
        localStorage.setItem("hasAuth", json)
    }
}

fun getAuthStatus(): HasAuth {
    val raw = localStorage.getItem("hasAuth") ?: return HasAuth(twitter = false)
    return try {
        Json.decodeFromString<HasAuth>(raw)
    } catch (e: Exception) {
        HasAuth(twitter = false)
    }
}

fun updateAuthStatus(transform: (HasAuth) -> HasAuth) {
    val current = getAuthStatus()
    setAuthStatus(transform(current))
}
